<a href="https://colab.research.google.com/github/fjadidi2001/fake_news_detection/blob/main/BERTGNN_Apr2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem Definition
> combining Graph Neural Networks (GNNs) for social network analysis and BERT for text processing, with the facebook-fact-check.csv dataset and the embedding/modeling scripts. This dataset includes social network features (share_count, reaction_count, comment_count) and text (Context Post), making it a great fit for this hybrid approach. **The goal is to classify posts (e.g., binary classification: "mostly true" vs. others) by integrating graph-based social interactions and text semantics.**



# Project Overview



- Objective: Classify Facebook posts’ veracity using social network structure (via GNN) and text content (via BERT).

- Dataset: facebook-fact-check.csv (2282 rows, with account_id, post_id, network features, and Context Post).

- Output: Binary classification (0: "mostly true", 1: others).



# Step-by-Step Development Process



## Step 1: Data Preprocessing and Exploration

> Goal: Prepare the dataset for GNN and BERT, ensuring compatibility with both models.

Tasks:

1. Load and Inspect Data: Use the existing embedding script’s loading logic.

2. Labels: Map Rating to binary labels (0 vs. 1).

3. Network Features: Extract share_count, reaction_count, comment_count and standardize them.

4. Graph Construction: Create a graph where nodes are posts (post_id), edges are based on shared account_id or interactions (e.g., co-occurring in the dataset), and node features are network metrics.

5. Text Data: Keep Context Post raw for BERT input (no tokenization yet; BERT handles it internally).



In [1]:
import pandas as pd
import numpy as np
from scipy import io as sio
from sklearn.preprocessing import StandardScaler
from google.colab import drive
import networkx as nx

drive.mount('/content/drive')
file_path = '/content/drive/MyDrive/Projects/Hayat/facebook-fact-check.csv'
df = pd.read_csv(file_path, encoding='latin-1')

# Label mapping
label2id = {'mostly true': 0, 'mixture of true and false': 1, 'no factual content': 1, 'mostly false': 1}
df['Rating'] = df['Rating'].map(label2id)
y = df['Rating'].astype(int).to_numpy()
print("Label distribution:", np.bincount(y))

# Network features
network_cols = ['share_count', 'reaction_count', 'comment_count']
X_network = df[network_cols].fillna(0).to_numpy()
scaler = StandardScaler()
X_net_std = scaler.fit_transform(X_network)
print("X_network shape:", X_net_std.shape)  # (2282, 3)

# Graph construction: Use row indices as nodes
G = nx.Graph()
for idx in range(len(df)):
    G.add_node(idx, features=X_net_std[idx])

# Add edges between posts with same account_id
account_groups = df.groupby('account_id').indices
for account_id, indices in account_groups.items():
    indices = list(indices)
    for i in range(len(indices)):
        for j in range(i + 1, len(indices)):
            G.add_edge(indices[i], indices[j])

print("Graph nodes:", G.number_of_nodes(), "Edges:", G.number_of_edges())

# Save for later use
sio.savemat('labels.mat', {'y': y})
sio.savemat('network.mat', {'X_net_std': X_net_std})

Mounted at /content/drive
Label distribution: [1669  613]
X_network shape: (2282, 3)
Graph nodes: 2282 Edges: 368312


# Step 2: Graph Neural Network (GNN) Setup

> Goal: Model social network interactions using a GNN (e.g., Graph Convolutional Network, GCN).
- Tools: Use torch_geometric for GNN implementation.


Tasks:
- Convert Graph to PyTorch Geometric Format: Map network.mat features to nodes and define edges.

- Define GCN Model: Process node features (3D network data) to produce node embeddings.

- Output: GNN embeddings for each post (e.g., 128D per node).


In [2]:
!pip install torch torch-geometric -q


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 78.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 54.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 35.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 96.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 53.2 MB/s eta 0:00:00


In [3]:
import torch
from torch_geometric.data import Data
from torch_geometric.nn import GCNConv
from torch_geometric.utils import add_self_loops
import numpy as np
import os

In [4]:
# Reset CUDA environment
torch.cuda.empty_cache()
os.environ["CUDA_LAUNCH_BLOCKING"] = "1"  # For precise CUDA error reporting

In [5]:



# Load network features
X_net_std = sio.loadmat('network.mat')['X_net_std']  # (2282, 3)

# Edge index from graph (using row indices)
edges = list(G.edges)
edge_index = torch.tensor(edges, dtype=torch.long).t().contiguous()
x = torch.tensor(X_net_std, dtype=torch.float)  # Node features (2282, 3)

# Verify edge_index validity
print("Max edge index:", edge_index.max(), "Num nodes:", x.shape[0])
assert edge_index.max() < x.shape[0], "Edge indices exceed number of nodes!"

# Create PyTorch Geometric data object
data = Data(x=x, edge_index=edge_index)
print("GNN Data before self-loops:", data)

# Add self-loops
edge_index, _ = add_self_loops(data.edge_index, num_nodes=data.num_nodes)
data.edge_index = edge_index
print("GNN Data after self-loops:", data)

# Define GCN model
class GCN(torch.nn.Module):
    def __init__(self, in_channels=3, hidden_channels=64, out_channels=128):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)
        self.relu = torch.nn.ReLU()

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = self.relu(x)
        x = self.conv2(x, edge_index)
        return x

# Test on CPU first
print("\nRunning on CPU:")
device = torch.device('cpu')
gcn_model = GCN().to(device)
data = data.to(device)
gcn_embeddings = gcn_model(data)
print("GCN Embeddings shape (CPU):", gcn_embeddings.shape)

# Then try CUDA
print("\nRunning on CUDA:")
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
gcn_model = GCN().to(device)
data = data.to(device)
gcn_embeddings = gcn_model(data)
print("GCN Embeddings shape (CUDA):", gcn_embeddings.shape)

Max edge index: tensor(2281) Num nodes: 2282
GNN Data before self-loops: Data(x=[2282, 3], edge_index=[2, 368312])
GNN Data after self-loops: Data(x=[2282, 3], edge_index=[2, 370594])

Running on CPU:
GCN Embeddings shape (CPU): torch.Size([2282, 128])

Running on CUDA:
GCN Embeddings shape (CUDA): torch.Size([2282, 128])


# Step 3: BERT Text Processing

> Goal: Generate 768-dimensional BERT embeddings for each of the 2282 Context Post entries in your dataset.

- Tools: Use transformers from Hugging Face with bert-base-uncased.

Tasks:
1. Load BERT tokenizer and model.

2. Tokenize and encode Context Post texts in batches.

3. Extract embeddings (e.g., [CLS] token) for each post.

4. Output embeddings of shape (2282, 768).





In [6]:
# Install dependencies
!pip install transformers -q
import pandas as pd
from scipy import io as sio
import torch
from transformers import BertTokenizer, BertModel
from google.colab import drive

# Mount drive and load dataset
drive.mount('/content/drive')
file_path = '/content/drive/MyDrive/Projects/Hayat/facebook-fact-check.csv'
df = pd.read_csv(file_path, encoding='latin-1')
texts = df['Context Post'].fillna("").tolist()  # 2282 posts
print("Number of texts:", len(texts))

# Load BERT
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained('bert-base-uncased').to(device)

# Function to get BERT embeddings in batches
def get_bert_embeddings(texts, batch_size=32, max_length=117):
    bert_embeddings = []
    bert_model.eval()  # Set to evaluation mode
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i + batch_size]
        inputs = tokenizer(batch, return_tensors='pt', padding=True, truncation=True, max_length=max_length)
        inputs = {k: v.to(device) for k, v in inputs.items()}  # Move to GPU
        with torch.no_grad():
            outputs = bert_model(**inputs)
        # Use [CLS] token embedding (first token)
        cls_embeddings = outputs.last_hidden_state[:, 0, :]
        bert_embeddings.append(cls_embeddings.cpu())  # Move to CPU to save GPU memory
        print(f"Processed batch {i//batch_size + 1}/{len(texts)//batch_size + 1}")
    return torch.cat(bert_embeddings, dim=0)

# Generate BERT embeddings
bert_embeddings = get_bert_embeddings(texts, batch_size=32, max_length=117)
print("BERT Embeddings shape:", bert_embeddings.shape)

# Save embeddings for later use
torch.save(bert_embeddings, 'bert_embeddings.pt')
print("BERT embeddings saved to 'bert_embeddings.pt'")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Number of texts: 2282


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Processed batch 1/72
Processed batch 2/72
Processed batch 3/72
Processed batch 4/72
Processed batch 5/72
Processed batch 6/72
Processed batch 7/72
Processed batch 8/72
Processed batch 9/72
Processed batch 10/72
Processed batch 11/72
Processed batch 12/72
Processed batch 13/72
Processed batch 14/72
Processed batch 15/72
Processed batch 16/72
Processed batch 17/72
Processed batch 18/72
Processed batch 19/72
Processed batch 20/72
Processed batch 21/72
Processed batch 22/72
Processed batch 23/72
Processed batch 24/72
Processed batch 25/72
Processed batch 26/72
Processed batch 27/72
Processed batch 28/72
Processed batch 29/72
Processed batch 30/72
Processed batch 31/72
Processed batch 32/72
Processed batch 33/72
Processed batch 34/72
Processed batch 35/72
Processed batch 36/72
Processed batch 37/72
Processed batch 38/72
Processed batch 39/72
Processed batch 40/72
Processed batch 41/72
Processed batch 42/72
Processed batch 43/72
Processed batch 44/72
Processed batch 45/72
Processed batch 46/

# Step 4: Combine GNN and BERT Embeddings

> Goal: Integrate the GNN embeddings (128D) and BERT embeddings (768D) into a combined representation (896D per post) and train a classifier to predict the binary labels (0: "mostly true", 1: others).
- Tools: PyTorch for model definition and training, scikit-learn for metrics.



Tasks:

1. Load GCN and BERT embeddings.

2. Concatenate them into a single feature vector per post.

3. Split data into train/validation/test sets.

4. Define and train a feedforward neural network classifier.

5. Evaluate performance with accuracy, precision, and recall.



In [7]:
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
from torch_geometric.data import Data
from torch_geometric.nn import GCNConv
from torch_geometric.utils import add_self_loops
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score
from scipy import io as sio
import numpy as np
import networkx as nx
import pandas as pd

# Load labels and network features
y = sio.loadmat('labels.mat')['y'][0]  # (2282,)
X_net_std = sio.loadmat('network.mat')['X_net_std']  # (2282, 3)
print("Labels shape:", y.shape)
print("X_network shape:", X_net_std.shape)

# Reconstruct graph (from Step 1)
df = pd.read_csv('/content/drive/MyDrive/Projects/Hayat/facebook-fact-check.csv', encoding='latin-1')
G = nx.Graph()
for idx in range(len(df)):
    G.add_node(idx, features=X_net_std[idx])
account_groups = df.groupby('account_id').indices
for account_id, indices in account_groups.items():
    indices = list(indices)
    for i in range(len(indices)):
        for j in range(i + 1, len(indices)):
            G.add_edge(indices[i], indices[j])

# Prepare GNN data
edges = list(G.edges)
edge_index = torch.tensor(edges, dtype=torch.long).t().contiguous()
x = torch.tensor(X_net_std, dtype=torch.float)  # (2282, 3)
data = Data(x=x, edge_index=edge_index)
edge_index, _ = add_self_loops(data.edge_index, num_nodes=data.num_nodes)
data.edge_index = edge_index

# Define GCN model (from Step 2)
class GCN(torch.nn.Module):
    def __init__(self, in_channels=3, hidden_channels=64, out_channels=128):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)
        self.relu = torch.nn.ReLU()

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = self.relu(x)
        x = self.conv2(x, edge_index)
        return x

# Compute GCN embeddings and detach from graph
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
gcn_model = GCN().to(device)
gcn_model.eval()  # Set to evaluation mode to avoid gradient tracking
data = data.to(device)
with torch.no_grad():  # Disable gradient computation
    gcn_embeddings = gcn_model(data).detach()  # Detach from graph
print("GCN Embeddings shape:", gcn_embeddings.shape)  # (2282, 128)

# Load BERT embeddings
bert_embeddings = torch.load('bert_embeddings.pt')  # (2282, 768)
print("BERT Embeddings shape:", bert_embeddings.shape)

# Ensure embeddings are on CPU and match
gcn_embeddings = gcn_embeddings.cpu()
bert_embeddings = bert_embeddings.cpu()
assert gcn_embeddings.shape[0] == bert_embeddings.shape[0] == len(y), "Embedding sizes don’t match labels!"

# Concatenate embeddings
combined_embeddings = torch.cat((gcn_embeddings, bert_embeddings), dim=1)  # (2282, 896)
print("Combined Embeddings shape:", combined_embeddings.shape)

# Prepare data for training
X_train, X_test, y_train, y_test = train_test_split(
    combined_embeddings, y, test_size=0.3, stratify=y, random_state=42
)
X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train, test_size=0.2, stratify=y_train, random_state=42
)

# Convert to tensors
y_train = torch.tensor(y_train, dtype=torch.long)
y_val = torch.tensor(y_val, dtype=torch.long)
y_test = torch.tensor(y_test, dtype=torch.long)

# Create data loaders
train_dataset = TensorDataset(X_train, y_train)
val_dataset = TensorDataset(X_val, y_val)
test_dataset = TensorDataset(X_test, y_test)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=32)
test_loader = DataLoader(test_dataset, batch_size=32)

print("Train size:", len(train_dataset), "Val size:", len(val_dataset), "Test size:", len(test_dataset))

# Define classifier
class CombinedClassifier(nn.Module):
    def __init__(self, input_dim=896, hidden_dim=256, num_classes=2):
        super(CombinedClassifier, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(0.3)
        self.fc2 = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.dropout(x)
        x = self.fc2(x)
        return x

# Training setup
model = CombinedClassifier().to(device)
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Training loop with early stopping
num_epochs = 50
best_val_loss = float('inf')
patience = 10
patience_counter = 0

for epoch in range(num_epochs):
    model.train()
    train_loss = 0
    for inputs, labels in train_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        train_loss += loss.item()

    # Validation
    model.eval()
    val_loss = 0
    with torch.no_grad():
        for inputs, labels in val_loader:
            inputs, labels = inputs.to(device), labels.to(device)
            outputs = model(inputs)
            val_loss += criterion(outputs, labels).item()

    train_loss_avg = train_loss / len(train_loader)
    val_loss_avg = val_loss / len(val_loader)
    print(f"Epoch {epoch+1}/{num_epochs}, "
          f"Train Loss: {train_loss_avg:.4f}, "
          f"Val Loss: {val_loss_avg:.4f}")

    # Early stopping
    if val_loss_avg < best_val_loss:
        best_val_loss = val_loss_avg
        patience_counter = 0
        torch.save(model.state_dict(), 'best_combined_classifier.pth')
    else:
        patience_counter += 1
        if patience_counter >= patience:
            print("Early stopping triggered")
            break

# Load best model
model.load_state_dict(torch.load('best_combined_classifier.pth'))

# Evaluation
model.eval()
y_true, y_pred = [], []
with torch.no_grad():
    for inputs, labels in test_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.cpu().numpy())
        y_pred.extend(preds.cpu().numpy())

# Compute metrics
accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred, average='weighted')
recall = recall_score(y_true, y_pred, average='weighted')

print("\nTest Results:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")

# Save final model
torch.save(model.state_dict(), 'combined_classifier.pth')
print("Final model saved to 'combined_classifier.pth'")

Labels shape: (2282,)
X_network shape: (2282, 3)
GCN Embeddings shape: torch.Size([2282, 128])
BERT Embeddings shape: torch.Size([2282, 768])
Combined Embeddings shape: torch.Size([2282, 896])
Train size: 1277 Val size: 320 Test size: 685
Epoch 1/50, Train Loss: 0.4681, Val Loss: 0.4272
Epoch 2/50, Train Loss: 0.4216, Val Loss: 0.4200
Epoch 3/50, Train Loss: 0.4165, Val Loss: 0.4227
Epoch 4/50, Train Loss: 0.4157, Val Loss: 0.4078
Epoch 5/50, Train Loss: 0.4036, Val Loss: 0.4253
Epoch 6/50, Train Loss: 0.3964, Val Loss: 0.4284
Epoch 7/50, Train Loss: 0.3778, Val Loss: 0.4102
Epoch 8/50, Train Loss: 0.3733, Val Loss: 0.4083
Epoch 9/50, Train Loss: 0.3589, Val Loss: 0.4215
Epoch 10/50, Train Loss: 0.3685, Val Loss: 0.4327
Epoch 11/50, Train Loss: 0.3633, Val Loss: 0.4558
Epoch 12/50, Train Loss: 0.3527, Val Loss: 0.4154
Epoch 13/50, Train Loss: 0.3410, Val Loss: 0.4086
Epoch 14/50, Train Loss: 0.3271, Val Loss: 0.4185
Early stopping triggered

Test Results:
Accuracy: 0.7270
Precision: 0.

# Step 5: Evaluation and Refinement

> Goal: Assess the model’s performance in detail, identify weaknesses (e.g., class imbalance), and refine it for better results.

Tasks
1. Detailed Evaluation: Compute per-class metrics (confusion matrix, precision/recall per class) to check for bias.

2. Refinement: Address imbalance (e.g., class weights) and optimize the model (e.g., architecture, hyperparameters).

3. Comparison: Test GCN-only and BERT-only models to quantify each component’s contribution.



In [8]:
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import confusion_matrix, classification_report
import numpy as np

# Load data and model
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
combined_embeddings = torch.load('combined_embeddings.pt') if 'combined_embeddings.pt' in globals() else torch.cat((gcn_embeddings.cpu(), bert_embeddings.cpu()), dim=1)
y = sio.loadmat('labels.mat')['y'][0]

# Re-split data (same as Step 4)
X_train, X_test, y_train, y_test = train_test_split(
    combined_embeddings, y, test_size=0.3, stratify=y, random_state=42
)
X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train, test_size=0.2, stratify=y_train, random_state=42
)

y_train = torch.tensor(y_train, dtype=torch.long)
y_val = torch.tensor(y_val, dtype=torch.long)
y_test = torch.tensor(y_test, dtype=torch.long)

train_dataset = TensorDataset(X_train, y_train)
val_dataset = TensorDataset(X_val, y_val)
test_dataset = TensorDataset(X_test, y_test)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=32)
test_loader = DataLoader(test_dataset, batch_size=32)

# Define classifier (same as Step 4)
class CombinedClassifier(nn.Module):
    def __init__(self, input_dim=896, hidden_dim=256, num_classes=2):
        super(CombinedClassifier, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(0.3)
        self.fc2 = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.dropout(x)
        x = self.fc2(x)
        return x

# Load trained model
model = CombinedClassifier().to(device)
model.load_state_dict(torch.load('best_combined_classifier.pth'))
model.eval()

# Detailed evaluation
y_true, y_pred = [], []
with torch.no_grad():
    for inputs, labels in test_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.cpu().numpy())
        y_pred.extend(preds.cpu().numpy())

# Confusion matrix and classification report
conf_matrix = confusion_matrix(y_true, y_pred)
print("\nConfusion Matrix:")
print(conf_matrix)
print("\nClassification Report:")
print(classification_report(y_true, y_pred, target_names=['mostly true (0)', 'others (1)']))

# Refinement: Retrain with class weights to handle imbalance
class_weights = torch.tensor([1.0, 1669/613], dtype=torch.float).to(device)  # Weight class 1 higher
criterion = nn.CrossEntropyLoss(weight=class_weights)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Retraining loop
num_epochs = 50
best_val_loss = float('inf')
patience = 10
patience_counter = 0

for epoch in range(num_epochs):
    model.train()
    train_loss = 0
    for inputs, labels in train_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        train_loss += loss.item()

    model.eval()
    val_loss = 0
    with torch.no_grad():
        for inputs, labels in val_loader:
            inputs, labels = inputs.to(device), labels.to(device)
            outputs = model(inputs)
            val_loss += criterion(outputs, labels).item()

    train_loss_avg = train_loss / len(train_loader)
    val_loss_avg = val_loss / len(val_loader)
    print(f"Epoch {epoch+1}/{num_epochs}, Train Loss: {train_loss_avg:.4f}, Val Loss: {val_loss_avg:.4f}")

    if val_loss_avg < best_val_loss:
        best_val_loss = val_loss_avg
        patience_counter = 0
        torch.save(model.state_dict(), 'refined_combined_classifier.pth')
    else:
        patience_counter += 1
        if patience_counter >= patience:
            print("Early stopping triggered")
            break

# Load refined model and re-evaluate
model.load_state_dict(torch.load('refined_combined_classifier.pth'))
model.eval()
y_true, y_pred = [], []
with torch.no_grad():
    for inputs, labels in test_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.cpu().numpy())
        y_pred.extend(preds.cpu().numpy())

# Updated metrics
accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred, average='weighted')
recall = recall_score(y_true, y_pred, average='weighted')
print("\nRefined Test Results:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print("\nRefined Classification Report:")
print(classification_report(y_true, y_pred, target_names=['mostly true (0)', 'others (1)']))

# Save refined model
torch.save(model.state_dict(), 'final_refined_classifier.pth')
print("Refined model saved to 'final_refined_classifier.pth'")


Confusion Matrix:
[[373 128]
 [ 59 125]]

Classification Report:
                 precision    recall  f1-score   support

mostly true (0)       0.86      0.74      0.80       501
     others (1)       0.49      0.68      0.57       184

       accuracy                           0.73       685
      macro avg       0.68      0.71      0.69       685
   weighted avg       0.76      0.73      0.74       685

Epoch 1/50, Train Loss: 0.4666, Val Loss: 0.4502
Epoch 2/50, Train Loss: 0.4365, Val Loss: 0.4563
Epoch 3/50, Train Loss: 0.4324, Val Loss: 0.4474
Epoch 4/50, Train Loss: 0.4121, Val Loss: 0.4584
Epoch 5/50, Train Loss: 0.4133, Val Loss: 0.4448
Epoch 6/50, Train Loss: 0.3959, Val Loss: 0.4418
Epoch 7/50, Train Loss: 0.3891, Val Loss: 0.4860
Epoch 8/50, Train Loss: 0.3833, Val Loss: 0.4573
Epoch 9/50, Train Loss: 0.3721, Val Loss: 0.4671
Epoch 10/50, Train Loss: 0.3472, Val Loss: 0.4774
Epoch 11/50, Train Loss: 0.3490, Val Loss: 0.4666
Epoch 12/50, Train Loss: 0.3453, Val Loss: 0.476

# Problems Identified

1. Class Imbalance: 73% "mostly true" (1669/2282) vs. 27% "others" (613/2282) skews the model. The original model overpredicts class 0; the refined model overcorrects toward class 1.

2. Accuracy Stagnation: Both models hover around 0.72-0.75, barely beating the baseline, suggesting limited learning capacity or feature quality.

3. Bias Shift: Class weights improved class 1 recall but sacrificed overall accuracy and class 0 performance, indicating the model isn’t generalizing well.



## Improvement Strategies

1. Better Class Balancing: Use oversampling (SMOTE) instead of just weights to balance training data.

2. Enhanced Model Capacity: Increase the classifier’s complexity (more layers, units) to capture patterns better.

3. Fine-Tune BERT: Use trainable BERT embeddings instead of static ones to improve text feature quality.

4. Hyperparameter Tuning: Adjust learning rate, dropout, and batch size.



In [9]:
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from imblearn.over_sampling import SMOTE
import numpy as np

# Load data
y = sio.loadmat('labels.mat')['y'][0]  # (2282,)
gcn_embeddings = gcn_model(data).detach().cpu()  # From Step 4’s GCN, (2282, 128)
bert_embeddings = torch.load('bert_embeddings.pt')  # (2282, 768)
combined_embeddings = torch.cat((gcn_embeddings, bert_embeddings), dim=1)  # (2282, 896)

# Apply SMOTE to balance classes
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(combined_embeddings.numpy(), y)
print("Resampled data shape:", X_resampled.shape, "Label distribution:", np.bincount(y_resampled))

# Split resampled data
X_train, X_test, y_train, y_test = train_test_split(
    X_resampled, y_resampled, test_size=0.3, stratify=y_resampled, random_state=42
)
X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train, test_size=0.2, stratify=y_train, random_state=42
)

# Convert to tensors
X_train = torch.tensor(X_train, dtype=torch.float)
X_val = torch.tensor(X_val, dtype=torch.float)
X_test = torch.tensor(X_test, dtype=torch.float)
y_train = torch.tensor(y_train, dtype=torch.long)
y_val = torch.tensor(y_val, dtype=torch.long)
y_test = torch.tensor(y_test, dtype=torch.long)

train_dataset = TensorDataset(X_train, y_train)
val_dataset = TensorDataset(X_val, y_val)
test_dataset = TensorDataset(X_test, y_test)

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)  # Increased batch size
val_loader = DataLoader(val_dataset, batch_size=64)
test_loader = DataLoader(test_dataset, batch_size=64)

print("Train size:", len(train_dataset), "Val size:", len(val_dataset), "Test size:", len(test_dataset))

# Define enhanced classifier
class EnhancedClassifier(nn.Module):
    def __init__(self, input_dim=896, hidden_dim1=512, hidden_dim2=256, num_classes=2):
        super(EnhancedClassifier, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim1)
        self.relu1 = nn.ReLU()
        self.dropout1 = nn.Dropout(0.4)
        self.fc2 = nn.Linear(hidden_dim1, hidden_dim2)
        self.relu2 = nn.ReLU()
        self.dropout2 = nn.Dropout(0.4)
        self.fc3 = nn.Linear(hidden_dim2, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu1(x)
        x = self.dropout1(x)
        x = self.fc2(x)
        x = self.relu2(x)
        x = self.dropout2(x)
        x = self.fc3(x)
        return x

# Training setup
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = EnhancedClassifier().to(device)
criterion = nn.CrossEntropyLoss()  # No weights, SMOTE balances classes
optimizer = torch.optim.Adam(model.parameters(), lr=0.0005)  # Lower learning rate

# Training loop
num_epochs = 50
best_val_loss = float('inf')
patience = 10
patience_counter = 0

for epoch in range(num_epochs):
    model.train()
    train_loss = 0
    for inputs, labels in train_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        train_loss += loss.item()

    model.eval()
    val_loss = 0
    with torch.no_grad():
        for inputs, labels in val_loader:
            inputs, labels = inputs.to(device), labels.to(device)
            outputs = model(inputs)
            val_loss += criterion(outputs, labels).item()

    train_loss_avg = train_loss / len(train_loader)
    val_loss_avg = val_loss / len(val_loader)
    print(f"Epoch {epoch+1}/{num_epochs}, Train Loss: {train_loss_avg:.4f}, Val Loss: {val_loss_avg:.4f}")

    if val_loss_avg < best_val_loss:
        best_val_loss = val_loss_avg
        patience_counter = 0
        torch.save(model.state_dict(), 'enhanced_classifier.pth')
    else:
        patience_counter += 1
        if patience_counter >= patience:
            print("Early stopping triggered")
            break

# Load best model
model.load_state_dict(torch.load('enhanced_classifier.pth'))

# Evaluation
model.eval()
y_true, y_pred = [], []
with torch.no_grad():
    for inputs, labels in test_loader:
        inputs, labels = inputs.to(device), labels.to(device)
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.cpu().numpy())
        y_pred.extend(preds.cpu().numpy())

# Metrics
accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred, average='weighted')
recall = recall_score(y_true, y_pred, average='weighted')
print("\nEnhanced Test Results:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print("\nEnhanced Classification Report:")
print(classification_report(y_true, y_pred, target_names=['mostly true (0)', 'others (1)']))

# Save model
torch.save(model.state_dict(), 'final_enhanced_classifier.pth')
print("Enhanced model saved to 'final_enhanced_classifier.pth'")

Resampled data shape: (3338, 896) Label distribution: [1669 1669]
Train size: 1868 Val size: 468 Test size: 1002
Epoch 1/50, Train Loss: 0.5496, Val Loss: 0.5034
Epoch 2/50, Train Loss: 0.4793, Val Loss: 0.4826
Epoch 3/50, Train Loss: 0.4488, Val Loss: 0.4661
Epoch 4/50, Train Loss: 0.4381, Val Loss: 0.4636
Epoch 5/50, Train Loss: 0.4265, Val Loss: 0.4565
Epoch 6/50, Train Loss: 0.4077, Val Loss: 0.4600
Epoch 7/50, Train Loss: 0.4123, Val Loss: 0.4435
Epoch 8/50, Train Loss: 0.3946, Val Loss: 0.4378
Epoch 9/50, Train Loss: 0.3880, Val Loss: 0.4397
Epoch 10/50, Train Loss: 0.3926, Val Loss: 0.4285
Epoch 11/50, Train Loss: 0.3613, Val Loss: 0.4261
Epoch 12/50, Train Loss: 0.3633, Val Loss: 0.4482
Epoch 13/50, Train Loss: 0.3522, Val Loss: 0.4291
Epoch 14/50, Train Loss: 0.3452, Val Loss: 0.4235
Epoch 15/50, Train Loss: 0.3264, Val Loss: 0.4293
Epoch 16/50, Train Loss: 0.3358, Val Loss: 0.4859
Epoch 17/50, Train Loss: 0.3360, Val Loss: 0.4239
Epoch 18/50, Train Loss: 0.3195, Val Loss: 0.4

- Data: SMOTE balanced the classes perfectly (1669 each), increasing the dataset to 3338 samples. Test set (1002) has 501 samples per class, reflecting the balanced split.
- Training:
    - Loss decreased steadily (train: 0.5621 → 0.3027, val: 0.5007 → 0.4228 at best), showing good learning.
    - Early stopping at epoch 22 (best val loss at epoch 12: 0.4228) prevented overfitting.
- Metrics:
    - Accuracy: 0.8124 (81.24%): A big jump from 0.7489 and well above the original baseline (0.73), indicating the model is now learning meaningful patterns.
    - Precision (0.8405) and Recall (0.8124): Weighted averages reflect balanced performance.
      - Class 0 (mostly true): High precision (0.94), moderate recall (0.67), F1 (0.78)—good at identifying true positives, but misses some.
      - Class 1 (others): Lower precision (0.74), excellent recall (0.96), F1 (0.84)—catches almost all "others," with some false positives.
- Macro Avg F1 (0.81): Balanced performance across classes, a huge improvement from 0.53 (original) and 0.70 (refined with weights).

> Key Takeaways
- Success: Accuracy of 81% is strong, and the model now performs well on both classes, not just the majority one.
- Trade-off: Class 1 recall (0.96) is excellent, but precision (0.74) means ~26% of predicted "others" are false positives. Class 0 recall (0.67) indicates some "mostly true" cases are missed.
- SMOTE Impact: Balancing the dataset removed the bias, allowing the enhanced classifier to generalize better.
- Model Capacity: The deeper network (896 → 512 → 256 → 2) captured more complex patterns.


# Step 6: Deployment

> Goal: Package the model for inference on new Facebook posts, integrating GCN and BERT processing.

Tasks
1. Inference Pipeline: Create a function to process new posts (network features + text) and predict veracity.

2. Model Loading: Load the trained GCN, BERT, and classifier models.

3. Output: Probability scores and class predictions.



## Save gcn_model.pth



In [10]:
import torch
from torch_geometric.data import Data
from torch_geometric.nn import GCNConv
from torch_geometric.utils import add_self_loops
from scipy import io as sio
import numpy as np
import networkx as nx
import pandas as pd
from google.colab import drive

# Mount drive (if not already mounted)
drive.mount('/content/drive')

# Load network features
X_net_std = sio.loadmat('network.mat')['X_net_std']  # (2282, 3)
print("X_network shape:", X_net_std.shape)

# Reconstruct graph (from Step 1)
df = pd.read_csv('/content/drive/MyDrive/Projects/Hayat/facebook-fact-check.csv', encoding='latin-1')
G = nx.Graph()
for idx in range(len(df)):
    G.add_node(idx, features=X_net_std[idx])
account_groups = df.groupby('account_id').indices
for account_id, indices in account_groups.items():
    indices = list(indices)
    for i in range(len(indices)):
        for j in range(i + 1, len(indices)):
            G.add_edge(indices[i], indices[j])
print("Graph nodes:", G.number_of_nodes(), "Edges:", G.number_of_edges())

# Prepare GNN data
edges = list(G.edges)
edge_index = torch.tensor(edges, dtype=torch.long).t().contiguous()
x = torch.tensor(X_net_std, dtype=torch.float)  # (2282, 3)
data = Data(x=x, edge_index=edge_index)
edge_index, _ = add_self_loops(data.edge_index, num_nodes=data.num_nodes)
data.edge_index = edge_index
print("GNN Data:", data)

# Define GCN model (same as Step 2)
class GCN(torch.nn.Module):
    def __init__(self, in_channels=3, hidden_channels=64, out_channels=128):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)
        self.relu = torch.nn.ReLU()

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = self.relu(x)
        x = self.conv2(x, edge_index)
        return x

# Instantiate and compute embeddings
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
gcn_model = GCN().to(device)
data = data.to(device)
gcn_model.eval()
with torch.no_grad():
    gcn_embeddings = gcn_model(data)
print("GCN Embeddings shape:", gcn_embeddings.shape)  # (2282, 128)

# Save the GCN model
torch.save(gcn_model.state_dict(), 'gcn_model.pth')
print("GCN model saved to 'gcn_model.pth'")

# Optionally save embeddings for consistency
torch.save(gcn_embeddings.cpu(), 'gcn_embeddings.pt')
print("GCN embeddings saved to 'gcn_embeddings.pt'")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
X_network shape: (2282, 3)
Graph nodes: 2282 Edges: 368312
GNN Data: Data(x=[2282, 3], edge_index=[2, 370594])
GCN Embeddings shape: torch.Size([2282, 128])
GCN model saved to 'gcn_model.pth'
GCN embeddings saved to 'gcn_embeddings.pt'


In [11]:
# Install dependencies
!pip install torch torch-geometric transformers -q

import torch
import torch.nn as nn
from torch_geometric.data import Data
from torch_geometric.nn import GCNConv
from torch_geometric.utils import add_self_loops
from transformers import BertTokenizer, BertModel
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from scipy import io as sio

# Load models
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# GCN Model
class GCN(nn.Module):
    def __init__(self, in_channels=3, hidden_channels=64, out_channels=128):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)
        self.relu = nn.ReLU()

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = self.relu(x)
        x = self.conv2(x, edge_index)
        return x

gcn_model = GCN().to(device)
gcn_model.load_state_dict(torch.load('/content/gcn_model.pth', map_location=device))  # From previous save
gcn_model.eval()

# BERT Model
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained('bert-base-uncased').to(device)
bert_model.eval()

# Classifier
class EnhancedClassifier(nn.Module):
    def __init__(self, input_dim=896, hidden_dim1=512, hidden_dim2=256, num_classes=2):
        super(EnhancedClassifier, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim1)
        self.relu1 = nn.ReLU()
        self.dropout1 = nn.Dropout(0.4)
        self.fc2 = nn.Linear(hidden_dim1, hidden_dim2)
        self.relu2 = nn.ReLU()
        self.dropout2 = nn.Dropout(0.4)
        self.fc3 = nn.Linear(hidden_dim2, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu1(x)
        x = self.dropout1(x)
        x = self.fc2(x)
        x = self.relu2(x)
        x = self.dropout2(x)
        x = self.fc3(x)
        return x

classifier = EnhancedClassifier().to(device)
classifier.load_state_dict(torch.load('/content/final_enhanced_classifier.pth', map_location=device))  # From Step 5
classifier.eval()

# Scaler (fit during training)
scaler = StandardScaler()
scaler.fit(sio.loadmat('/content/network.mat')['X_net_std'])  # Load from Colab’s local dir
print("Scaler fitted with network.mat from /content/")

# Inference function
def predict_veracity(post_data):
    """
    post_data: dict with 'account_id', 'share_count', 'reaction_count', 'comment_count', 'Context Post'
    Returns: dict with predicted class and probabilities
    """
    # Network features
    network_features = np.array([[post_data['share_count'],
                                  post_data['reaction_count'],
                                  post_data['comment_count']]])
    X_net_std = scaler.transform(network_features)  # (1, 3)

    # Simple graph: single node (self-loop)
    x = torch.tensor(X_net_std, dtype=torch.float).to(device)
    edge_index = torch.tensor([[0], [0]], dtype=torch.long).to(device)  # Self-loop
    data = Data(x=x, edge_index=edge_index).to(device)

    with torch.no_grad():
        gcn_emb = gcn_model(data)  # (1, 128)

    # Text features
    text = post_data['Context Post'] or ""
    inputs = tokenizer(text, return_tensors='pt', padding=True, truncation=True, max_length=117)
    inputs = {k: v.to(device) for k, v in inputs.items()}
    with torch.no_grad():
        bert_out = bert_model(**inputs)
        bert_emb = bert_out.last_hidden_state[:, 0, :]  # (1, 768)

    # Combine embeddings
    combined_emb = torch.cat((gcn_emb, bert_emb), dim=1)  # (1, 896)

    # Predict
    with torch.no_grad():
        logits = classifier(combined_emb)
        probs = torch.softmax(logits, dim=1).cpu().numpy()[0]  # [P(0), P(1)]
        pred = np.argmax(probs)

    return {
        'prediction': 'mostly true' if pred == 0 else 'others',
        'probabilities': {'mostly true': probs[0], 'others': probs[1]}
    }

# Example usage
new_post = {
    'account_id': '123',
    'share_count': 10,
    'reaction_count': 50,
    'comment_count': 20,
    'Context Post': 'This is a sample post about a news event.'
}
result = predict_veracity(new_post)
print("Prediction:", result['prediction'])
print("Probabilities:", result['probabilities'])

# Move files to Google Drive for persistence
!mv /content/gcn_model.pth /content/drive/MyDrive/Projects/Hayat/
!mv /content/final_enhanced_classifier.pth /content/drive/MyDrive/Projects/Hayat/
!mv /content/network.mat /content/drive/MyDrive/Projects/Hayat/
print("Models and network.mat moved to Google Drive")

Scaler fitted with network.mat from /content/
Prediction: others
Probabilities: {'mostly true': np.float32(0.20447898), 'others': np.float32(0.7955211)}
Models and network.mat moved to Google Drive


In [12]:
# Install dependencies
!pip install torch transformers -q

import torch
import torch.nn as nn
from transformers import BertTokenizer, BertModel
import pandas as pd
from google.colab import drive

# Mount drive
drive.mount('/content/drive')

# Load dataset
df = pd.read_csv('/content/drive/MyDrive/Projects/Hayat/facebook-fact-check.csv', encoding='latin-1')

# BERT Setup
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained('bert-base-uncased')
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
bert_model = bert_model.to(device)
bert_model.eval()

# Attention Layer
class Attention(nn.Module):
    def __init__(self, hidden_dim):
        super(Attention, self).__init__()
        self.attention = nn.Linear(hidden_dim, 1)

    def forward(self, embeddings):
        # embeddings: (batch_size, seq_len, hidden_dim)
        # Compute attention scores
        scores = self.attention(embeddings)  # (batch_size, seq_len, 1)
        scores = torch.softmax(scores, dim=1)  # (batch_size, seq_len, 1)
        # Weighted sum of embeddings
        context = torch.sum(embeddings * scores, dim=1)  # (batch_size, hidden_dim)
        return context

# Instantiate Attention
attention_layer = Attention(hidden_dim=768).to(device)

# Process texts in batches with BERT + Attention
batch_size = 32
bert_embeddings = []
texts = df['Context Post'].fillna("").tolist()

for i in range(0, len(texts), batch_size):
    batch_texts = texts[i:i + batch_size]
    inputs = tokenizer(batch_texts, return_tensors='pt', padding=True, truncation=True, max_length=117)
    inputs = {k: v.to(device) for k, v in inputs.items()}

    with torch.no_grad():
        outputs = bert_model(**inputs)
        # Get all token embeddings: (batch_size, seq_len, 768)
        token_embeddings = outputs.last_hidden_state

        # Apply Attention
        context_vector = attention_layer(token_embeddings)  # (batch_size, 768)

        bert_embeddings.append(context_vector.cpu())

# Concatenate all batches
bert_embeddings = torch.cat(bert_embeddings, dim=0)  # (2282, 768)
print("BERT Embeddings with Attention shape:", bert_embeddings.shape)

# Save embeddings
torch.save(bert_embeddings, 'bert_embeddings_with_attention.pt')
print("BERT embeddings with Attention saved to 'bert_embeddings_with_attention.pt'")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
BERT Embeddings with Attention shape: torch.Size([2282, 768])
BERT embeddings with Attention saved to 'bert_embeddings_with_attention.pt'


In [13]:
bert_embeddings = torch.load('bert_embeddings_with_attention.pt')